<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item">

<li><span><a href="#For category prediction" data-toc-modified-id="For category prediction-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Category Prediction</a></span></li><ul class="toc-item">

<li><span><a href="#ItemKNN and SKNN"
data-toc-modified-id="ItemKNN and SKNN-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>ItemKNN and SKNN</a></span></li></ul>

<li><span><a href="#For category prediction" data-toc-modified-id="For category prediction-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Article prediction</a></span><ul class="toc-item">

<li><span><a href="#ItemKNN and SKNN"
data-toc-modified-id="ItemKNN and SKNN-1.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>ItemKNN and SKNN</a></span></li></ul>


<li><span><a href="#For category prediction" data-toc-modified-id="For category prediction-2"><span class="toc-item-num">3&nbsp;&nbsp;</span>GRU (loss type: TOP1)</a></span><ul class="toc-item">

<li><span><a href="#ItemKNN and SKNN"
data-toc-modified-id="ItemKNN and SKNN-1.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Change loss type (BPR loss)</a></span></li>
<li><span><a href="#ItemKNN and SKNN"
data-toc-modified-id="ItemKNN and SKNN-1.1"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Change loss type (Cross-Entropy loss)</a></span></li></ul>

</ul></div>

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&response_type=code&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly

Enter your authorization code:
··········
Mounted at /content/drive


In [0]:
import pandas as pd
import numpy as np
path="drive/My\ Drive/1001_project_Total"

# For category prediction

In [3]:
!python $path/Data/preprocess.py --Path $path/Data/day_one_category_region_device.csv --Feature True --Try True

Full train set
	Events: 110918
	Sessions: 42611
	Items: 37
Test set
	Events: 38097
	Sessions: 13790
	Items: 35
Train set
	Events: 82635
	Sessions: 31990
	Items: 37
Validation set
	Events: 28283
	Sessions: 10621
	Items: 32


## ItemKNN and SKNN

In [4]:
path="drive/My\ Drive/1001_project_Total"
!python $path/baseline/run.py --Path $path --Train day_one_train_full.txt --Test day_one_test_full.txt --Prediction category

100% 38097/38097 [00:45<00:00, 835.67it/s]
The accuracy of Random Prediction:
Recall@20: 0.24495001439914427
MRR@20: 0.07674378208380739


100% 38097/38097 [01:00<00:00, 633.41it/s]
The accuracy of item k-nearest neighbors:
Recall@20: 0.6048463405603324
MRR@20: 0.34281076194000454


100% 38097/38097 [07:15<00:00, 87.57it/s]
The accuracy of session k-nearest neighbors:
Recall@20: 0.7844653803431111
MRR@20: 0.4797509013342064


100% 38097/38097 [08:31<00:00, 74.46it/s]
The accuracy of session k-nearest neighbors with feature region:
Recall@20: 0.8906899247130456
MRR@20: 0.5167453068911785


100% 38097/38097 [08:40<00:00, 73.18it/s]
The accuracy of session k-nearest neighbors with features region and devive type:
Recall@20: 0.8902785205907763
MRR@20: 0.5167331443486748


Random Prediction:      
Recall@20: 0.24495001439914427    
MRR@20: 0.07674378208380739

ItemKNN:   
Recall@20: 0.6048463405603324  
MRR@20: 0.34281076194000454


SKNN (no other features):     
Recall@20: 0.7844653803431111   
MRR@20: 0.4797509013342064


SKNN (with region features):     
Recall@20: 0.8906899247130456  
MRR@20: 0.5167453068911785


SKNN (with regions and devices):   
Recall@20: 0.8902785205907763   
MRR@20: 0.5167331443486748

## Sequential Rule for Category

In [0]:
path="drive/My\ Drive/1001_project_total"
!python $path/baseline/SR_Cate.py

Data is loaded in: day_one_category_region_device.csv
The length of final data_session: 149071
# interactions on Train set x: 119146
# interactions on Test set x: 29925
# interactions on Train set y: 74009
# interactions on Test set y: 18641
(415, 7)
the hit rate for top 10 is: 0.9761278901346494
the MRR for top 10 is: 0.8650993330114621


# For article prediction

In [10]:
path="drive/My\ Drive/1001_project_Total"
!python $path/Data/preprocess.py --Path $path/Data/day_one_article_region_device.csv --Feature True --Try True

Full train set
	Events: 109760
	Sessions: 42235
	Items: 182
Test set
	Events: 19767
	Sessions: 7200
	Items: 135
Train set
	Events: 81784
	Sessions: 31704
	Items: 171
Validation set
	Events: 20487
	Sessions: 7854
	Items: 126


## ItemKNN and SKNN

In [12]:
path="drive/My\ Drive/1001_project_Total"
!python $path/baseline/run.py --Path $path --Train $path/Data/day_one_train_full.txt --Test  $path/Data/day_one_test_full.txt --Prediction article

100% 19767/19767 [00:36<00:00, 543.36it/s]
The accuracy of Random Prediction:
Recall@20: 0.04758494469642715
MRR@20: 0.015170843011616449


100% 19767/19767 [00:48<00:00, 406.53it/s]
The accuracy of item k-nearest neighbors:
Recall@20: 0.5290045356887085
MRR@20: 0.23028544777769208


100% 19767/19767 [01:57<00:00, 160.05it/s]
The accuracy of session k-nearest neighbors:
Recall@20: 0.6117609612477123
MRR@20: 0.2550600211438149


100% 19767/19767 [02:44<00:00, 119.99it/s]
The accuracy of session k-nearest neighbors with feature region:
Recall@20: 0.6395321079016472
MRR@20: 0.2620798551509978


100% 19767/19767 [02:44<00:00, 119.83it/s]
The accuracy of session k-nearest neighbors with features region and devive type:
Recall@20: 0.6392933874433039
MRR@20: 0.262155702577048


Random Prediction:      
Recall@20: 0.04758494469642715   
MRR@20: 0.015170843011616449

ItemKNN:   
Recall@20: 0.5290045356887085   
MRR@20: 0.23028544777769208


SKNN (no other features):     
Recall@20: 0.6117609612477123   
MRR@20: 0.2550600211438149


SKNN (with region features):     
Recall@20: 0.6395321079016472   
MRR@20: 0.2620798551509978


SKNN (with regions and devices):   
Recall@20: 0.6392933874433039   
MRR@20: 0.262155702577048



## Sequence Rule for Article

In [0]:

path="drive/My\ Drive/1001_project_total"
!python $path/baseline/SR_Art.py



Data is loaded in: day_one_article_region_device.csv
The length of final data_session: 149071
# interactions on Train set x: 119146
# interactions on Test set x: 29925
# interactions on Train set y: 74009
# interactions on Test set y: 18641
(6389, 7)
the hit rate for top 20 is: 0.9050480124456842
the MRR for top 20 is: 0.4841767472017995


# GRU (loss type: TOP1)



lr=0.01
hidden_size=500
test recall=0.615

lr=0.02
hidden_size=500
recall=0.621

lr=0.03
hidden_size=500
recall=0.621

lr=0.01
hidden_size=750, 
recall=0.629

lr=0.02
hidden_size=750, 
recall=0.628

lr=0.03
hidden_size=750, 
recall=0.604

lr=0.01
hidden_size=1000, 
recall=0.614

lr=0.02
hidden_size=1000, 
recall=0.620


  

In [0]:
lr=[0.01,0.02,0.03]
hidden_size=[500,750,1000]

In [9]:
import sys
sys.path.append('drive/My Drive/1001_project_Total/GRU4REC')
from pathlib import Path
import pandas as pd
import numpy as np
import torch
from modules.data import SessionDataset
from modules.model import GRU4REC


train = "drive/My Drive/1001_project_Total/Data/"+'day_one_train_full.txt'
test = "drive/My Drive/1001_project_Total/Data/"+'day_one_test_full.txt'

train_dataset = SessionDataset(train)
test_dataset = SessionDataset(test, itemmap=train_dataset.itemmap)
print("Loss type is TOP1: \n")
loss_type = 'TOP1'
def run(hidd_size,l,loss_type):
  for hidden_size in hidd_size:
    for lr in l:
      input_size = len(train_dataset.items)
      num_layers =1
      output_size = input_size
      batch_size = 50

      optimizer_type = 'Adagrad'
      weight_decay = 0
      momentum = 0
      eps = 1e-6

      n_epochs = 2
      use_cuda = True

      torch.manual_seed(7)

      model = GRU4REC(input_size, hidden_size, output_size,
                      num_layers=num_layers,
                      batch_size=batch_size,
                      optimizer_type=optimizer_type,
                      lr=lr,
                      weight_decay=weight_decay,
                      momentum=momentum,
                      eps=eps,
                      loss_type=loss_type,
                      use_cuda=use_cuda)
      model_name = 'GRU4REC'
      print("hidden size is {0}, lr is {1}".format(hidden_size,lr))
      model.train(train_dataset, n_epochs=n_epochs, model_name=model_name, save=True)
      model.test(test_dataset)
run(hidden_size,lr,loss_type)

/usr/local/lib/python3.6/dist-packages/IPython/core/interactiveshell.py:2822: DtypeWarning: Columns (0,1,2,3,4,5,6,7) have mixed types. Specify dtype option on import or set low_memory=False.
  if self.run_code(code, result):


Loss type is TOP1: 

hidden size is 500, lr is 0.01
Training GRU4REC...


/usr/local/lib/python3.6/dist-packages/torch/nn/functional.py:1351: UserWarning: nn.functional.sigmoid is deprecated. Use torch.sigmoid instead.
  warnings.warn("nn.functional.sigmoid is deprecated. Use torch.sigmoid instead.")


epoch: 1/loss:0.935/recall:0.719/mrr:0.342/time:0.106
epoch: 1/loss:0.922/recall:0.722/mrr:0.344/time:0.107
['loss:1.039', 'recall:0.615', 'mrr:0.270', 'time:0.008']
hidden size is 500, lr is 0.02
Training GRU4REC...
epoch: 1/loss:0.929/recall:0.706/mrr:0.334/time:0.105
The end of training~
['loss:1.039', 'recall:0.613', 'mrr:0.273', 'time:0.008']
hidden size is 500, lr is 0.03
Training GRU4REC...
epoch: 1/loss:0.928/recall:0.693/mrr:0.324/time:0.104
The end of training~
['loss:1.038', 'recall:0.602', 'mrr:0.271', 'time:0.008']
hidden size is 750, lr is 0.01
Training GRU4REC...
epoch: 1/loss:0.933/recall:0.724/mrr:0.342/time:0.115
The end of training~
['loss:1.041', 'recall:0.623', 'mrr:0.269', 'time:0.008']
hidden size is 750, lr is 0.02
Training GRU4REC...
epoch: 1/loss:0.929/recall:0.706/mrr:0.332/time:0.114
The end of training~
['loss:1.040', 'recall:0.607', 'mrr:0.270', 'time:0.008']
hidden size is 750, lr is 0.03
Training GRU4REC...
epoch: 1/loss:0.929/recall:0.693/mrr:0.321/time

## 2. Change loss type(BPR Loss)

In [10]:
loss_type = 'BPR'
run(hidden_size,lr,loss_type)

hidden size is 500, lr is 0.01
Training GRU4REC...
epoch: 1/loss:0.499/recall:0.716/mrr:0.345/time:0.098
epoch: 1/loss:0.455/recall:0.753/mrr:0.356/time:0.096
['loss:0.689', 'recall:0.607', 'mrr:0.266', 'time:0.008']
hidden size is 500, lr is 0.02
Training GRU4REC...
epoch: 1/loss:0.474/recall:0.717/mrr:0.338/time:0.096
epoch: 1/loss:0.442/recall:0.746/mrr:0.353/time:0.097
['loss:0.686', 'recall:0.615', 'mrr:0.271', 'time:0.008']
hidden size is 500, lr is 0.03
Training GRU4REC...
epoch: 1/loss:0.467/recall:0.712/mrr:0.330/time:0.096
epoch: 1/loss:0.439/recall:0.741/mrr:0.343/time:0.096
['loss:0.687', 'recall:0.614', 'mrr:0.266', 'time:0.008']
hidden size is 750, lr is 0.01
Training GRU4REC...
epoch: 1/loss:0.491/recall:0.720/mrr:0.345/time:0.109
epoch: 1/loss:0.450/recall:0.754/mrr:0.353/time:0.107
['loss:0.687', 'recall:0.617', 'mrr:0.267', 'time:0.008']
hidden size is 750, lr is 0.02
Training GRU4REC...
epoch: 1/loss:0.473/recall:0.720/mrr:0.337/time:0.108
epoch: 1/loss:0.442/recall:

## 3. Change loss type(Cross-Entropy loss)

In [11]:
loss_type = 'CrossEntropy'
run(hidden_size,lr,loss_type)

hidden size is 500, lr is 0.01
Training GRU4REC...
epoch: 1/loss:3.472/recall:0.670/mrr:0.330/time:0.087
epoch: 1/loss:3.353/recall:0.708/mrr:0.345/time:0.087
['loss:3.902', 'recall:0.591', 'mrr:0.267', 'time:0.008']
hidden size is 500, lr is 0.02
Training GRU4REC...
epoch: 1/loss:3.400/recall:0.667/mrr:0.328/time:0.090
epoch: 1/loss:3.314/recall:0.700/mrr:0.341/time:0.089
['loss:3.896', 'recall:0.585', 'mrr:0.267', 'time:0.008']
hidden size is 500, lr is 0.03
Training GRU4REC...
epoch: 1/loss:3.382/recall:0.660/mrr:0.321/time:0.089
epoch: 1/loss:3.304/recall:0.697/mrr:0.336/time:0.089
['loss:3.890', 'recall:0.577', 'mrr:0.264', 'time:0.008']
hidden size is 750, lr is 0.01
Training GRU4REC...
epoch: 1/loss:3.453/recall:0.676/mrr:0.333/time:0.105
epoch: 1/loss:3.340/recall:0.708/mrr:0.345/time:0.105
['loss:3.901', 'recall:0.595', 'mrr:0.269', 'time:0.008']
hidden size is 750, lr is 0.02
Training GRU4REC...
epoch: 1/loss:3.400/recall:0.674/mrr:0.327/time:0.104
epoch: 1/loss:3.312/recall: